File: terminal/bridge/lib/log.dart
import 'dart:convert';
import 'dart:io';

const _proc = 'bridge';

enum LogLevel {
  debug(0),
  info(1),
  warn(2),
  error(3);

  const LogLevel(this.priority);
  final int priority;

  static LogLevel fromEnv([String? raw]) {
    final value = (raw ?? Platform.environment['COMSTAR_LOG'] ?? 'info')
        .trim()
        .toLowerCase();
    return switch (value) {
      'debug' => LogLevel.debug,
      'warn' => LogLevel.warn,
      'error' => LogLevel.error,
      _ => LogLevel.info,
    };
  }
}

LogLevel _globalLevel = LogLevel.fromEnv();

void configureLogLevel(LogLevel level) => _globalLevel = level;

void logEvent({
  required LogLevel level,
  required String evt,
  required String msg,
  String? turnId,
  Map<String, dynamic>? data,
}) {
  if (level.priority < _globalLevel.priority) return;
  final entry = <String, dynamic>{
    'ts': DateTime.now().millisecondsSinceEpoch,
    'level': level.name,
    'proc': _proc,
    'evt': evt,
    'msg': msg,
  };
  if (turnId != null) entry['turn_id'] = turnId;
  if (data != null && data.isNotEmpty) entry['data'] = data;
  stdout.writeln(jsonEncode(entry));
}

void logDebug(String evt, String msg,
        {String? turnId, Map<String, dynamic>? data}) =>
    logEvent(
        level: LogLevel.debug,
        evt: evt,
        msg: msg,
        turnId: turnId,
        data: data);

void logInfo(String evt, String msg,
        {String? turnId, Map<String, dynamic>? data}) =>
    logEvent(
        level: LogLevel.info,
        evt: evt,
        msg: msg,
        turnId: turnId,
        data: data);

void logWarn(String evt, String msg,
        {String? turnId, Map<String, dynamic>? data}) =>
    logEvent(
        level: LogLevel.warn,
        evt: evt,
        msg: msg,
        turnId: turnId,
        data: data);

void logError(String evt, String msg,
        {String? turnId, Map<String, dynamic>? data}) =>
    logEvent(
        level: LogLevel.error,
        evt: evt,
        msg: msg,
        turnId: turnId,
        data: data);

class Span {
  Span(this.name, {this.turnId}) : _start = DateTime.now();

  final String name;
  final String? turnId;
  final DateTime _start;
  bool _closed = false;

  int get elapsedMs => DateTime.now().difference(_start).inMilliseconds;

  void close({Map<String, dynamic>? data}) {
    if (_closed) return;
    _closed = true;
    logInfo('span', '$name completed', turnId: turnId, data: {
      'name': name,
      'ms': elapsedMs,
      if (data != null) ...data,
    });
  }
}

File: terminal/audio/log.py
"""Structured JSON logging for the COMSTAR audio process."""

from __future__ import annotations

import json
import os
import sys
import time
from typing import Any

_PROC = "audio"
_LEVELS = {"debug": 0, "info": 1, "warn": 2, "error": 3}
_LEVEL = os.environ.get("COMSTAR_LOG", "info").strip().lower()


def _should_log(level: str) -> bool:
    return _LEVELS.get(level, 1) >= _LEVELS.get(_LEVEL, 1)


def log_event(
    level: str,
    evt: str,
    msg: str,
    *,
    turn_id: str | None = None,
    data: dict[str, Any] | None = None,
) -> None:
    if not _should_log(level):
        return
    entry: dict[str, Any] = {
        "ts": int(time.time() * 1000),
        "level": level,
        "proc": _PROC,
        "evt": evt,
        "msg": msg,
    }
    if turn_id is not None:
        entry["turn_id"] = turn_id
    if data:
        entry["data"] = data
    sys.stdout.write(json.dumps(entry, separators=(",", ":")) + "\n")
    sys.stdout.flush()


def log_info(evt: str, msg: str, **kwargs: Any) -> None:
    log_event("info", evt, msg, **kwargs)


def log_warn(evt: str, msg: str, **kwargs: Any) -> None:
    log_event("warn", evt, msg, **kwargs)


def log_error(evt: str, msg: str, **kwargs: Any) -> None:
    log_event("error", evt, msg, **kwargs)


class Span:
    def __init__(self, name: str, *, turn_id: str | None = None) -> None:
        self.name = name
        self.turn_id = turn_id
        self._start = time.monotonic()
        self._closed = False

    def close(self, **data: Any) -> None:
        if self._closed:
            return
        self._closed = True
        ms = int((time.monotonic() - self._start) * 1000)
        payload = {"name": self.name, "ms": ms, **data}
        log_info("span", f"{self.name} completed", turn_id=self.turn_id, data=payload)
